### PR TITLE
feat: Proper CSS api & css bundle

### PR DESCRIPTION
**Pull request by Vendicated**
Need to figure out a good way to do this in the userscript and also let fridge know he needs to update the loader

### CODE REVIEW

**Comment by Vendicated** [comment]
Issue: The updater will not download renderer.css. Need to find a fix for this / implement a temporary hack that will download the renderer.css file if it doesn't exist

File: src/api/Styles.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+    const style = ensureStyle(name);
+    style.vars = vars;
+    if (style.mount?.isConnected) compileStyle(style);
+};
+
+export const compileStyle = (style: Style) => {
+    if (!style.mount) throw new Error("Style is not mounted");
+
+    style.mount.textContent = style.source
+        .replace(/(?<=#\{)(\w+)(?=\})/g, (_, name) => style.vars[name] ?? "");
```
Comment (Vendicated) [inline src/api/Styles.ts:69]: what exactly does this do?

Comment (Vendicated) [reply src/api/Styles.ts:69]: it seems to enable some sort of variables? That should be documented

File: src/api/Styles.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+ *
+ * // Inside some plugin method like "start()" or "[option].onChange()"
+ * enableStyle(pluginStyle);
+ */
+export const enableStyle = (name: string) => {
+    const style = ensureStyle(name);
+
+    if (style.dom?.isConnected) return false;
+
+    style.dom ??= document.createElement("style");
```
Comment (Vendicated) [inline src/api/Styles.ts:46]: imo this should have some id so it's obvious where it comes from

Comment (Vendicated) [reply src/api/Styles.ts:46]: like id = "vc-styles-NAME"

Comment (Vap0r1ze) [reply src/api/Styles.ts:46]: true

File: src/api/Styles.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
+
+import type { MapValue } from "type-fest/source/entry";
+
+export type Style = MapValue<typeof VencordStyles>;
+
+export const styleMap = window.VencordStyles ??= new Map();
+
+export const ensureStyle = (name: string) => {
```
Comment (Vendicated) [inline src/api/Styles.ts:25]: ~~bikeshedding but~~ requireStyle would probably be a better name

File: src/api/Styles.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+ * setStyleClassNames(pluginStyle, classNames);
+ * enableStyle(pluginStyle);
+ * ```
+ * ```scss
+ * // -- plugin.css --
+ * .plugin-root [--thin]::-webkit-scrollbar { ... }
+ * ```
+ * ```scss
+ * // -- final stylesheet --
+ * .plugin-root .thin-31rlnD.scrollerBase-_bVAAt::-webkit-scrollbar { ... }
```
Comment (Vendicated) [inline src/api/Styles.ts:102]: pog

File: src/api/Styles.ts (hunk 3 above)
Comment (Vap0r1ze) [reply src/api/Styles.ts:25]: 🚴 F2 it u wont

File: scripts/build/common.mjs [hunk 5]
```diff
@@ -147,6 +147,31 @@ export const fileIncludePlugin = {
     }
 };
 
+const styleModule = await readFile("./scripts/build/module/style.js", "utf-8");
```
Comment (Vendicated) [inline scripts/build/common.mjs:150]:
```suggestion
const styleModule = readFileSync("./scripts/build/module/style.js", "utf-8");
```

sync is way faster than async and the await makes it synchronous anyway

Comment (Vap0r1ze) [reply scripts/build/common.mjs:150]: i figured it would allow esbuild to read multiple files in parallel

File: scripts/build/common.mjs [hunk 6]
```diff
@@ -147,6 +147,31 @@ export const fileIncludePlugin = {
...
+/**
+ * @type {import("esbuild").Plugin}
+ */
+export const stylePlugin = {
+    name: "style-plugin",
+    setup: ({ onResolve, onLoad }) => {
+        onResolve({ filter: /\.css\?managed$/, namespace: "file" }, ({ path, resolveDir }) => ({
+            path: relative(process.cwd(), join(resolveDir, path.replace("?managed", ""))),
```
Comment (Vap0r1ze) [inline scripts/build/common.mjs:158]: i willl figure this out later

**Comment by Vendicated** [comment]
This is basically ready now, just need to wait for ArmCord to support it

**Pull request merged**